### PR TITLE
Allow specifying of enabled protocols for `SchannelCredBuilder`.

### PR DESCRIPTION
**Pull request by frewsxcv**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -984,6 +1032,35 @@ mod test {
...
+        let creds = SchannelCredBuilder::new()
+            .enabled_protocols(&[Protocol::Ssl3])
+            .acquire(Direction::Outbound)
+            .unwrap();
+        let stream = TcpStream::connect("google.com:443").unwrap();
+        let stream = TlsStreamBuilder::new()
+            .domain("google.com")
+            .initialize(creds, stream);
+        assert_eq!(format!("{:?}", stream.err().unwrap()),
+                   "Error { repr: Custom(Custom { kind: Other, error: Error { code: 0x80090302, message: \"The function requested is not supported\" } }) }");
```
Comment (frewsxcv) [inline src/lib.rs]: Not sure of the best way to test this since `initialize` returns an `io::Error`

File: src/lib.rs [hunk 2]
```diff
@@ -984,6 +1032,35 @@ mod test {
...
+    }
+
+    #[test]
+    fn valid_protocol() {
+        // How to make this pass...
+        let creds = SchannelCredBuilder::new()
+            .enabled_protocols(&[Protocol::Tls10, Protocol::Tls11, Protocol::Tls12])
+            .supported_algorithms(&[Algorithm::RsaKeyx, Algorithm::Aes128, Algorithm::Sha1, Algorithm::Aes, Algorithm::Aes192, Algorithm::Aes256, Algorithm::Ecdh, Algorithm::RsaSign, Algorithm::Sha384, Algorithm::Sha512])
+            .acquire(Direction::Outbound);
+        assert_eq!(creds.err().unwrap().0, winapi::SEC_E_ALGORITHM_MISMATCH as winapi::DWORD);
```
Comment (frewsxcv) [inline src/lib.rs]: For some reason, I can't seem to get `enabled_algorithms` to _not_ return a `SEC_E_ALGORITHM_MISMATCH`

File: src/lib.rs [hunk 3]
```diff
@@ -188,13 +189,49 @@ pub enum Algorithm {
...
+                Protocol::Tls10 => winapi::SP_PROT_TLS1_0_SERVER,
+                Protocol::Tls11 => winapi::SP_PROT_TLS1_1_SERVER,
+                Protocol::Tls12 => winapi::SP_PROT_TLS1_2_SERVER,
+            }
+        } else {
+            match self {
+                Protocol::Ssl3 => winapi::SP_PROT_SSL3_CLIENT,
+                Protocol::Tls10 => winapi::SP_PROT_TLS1_0_SERVER,
+                Protocol::Tls11 => winapi::SP_PROT_TLS1_1_SERVER,
+                Protocol::Tls12 => winapi::SP_PROT_TLS1_2_SERVER,
```
Comment (frewsxcv) [inline src/lib.rs]: Oh, these should be `CLIENT`. Copy+paste fail

**Comment by frewsxcv** [comment]
I can squash before merging, just let me know

File: src/lib.rs [hunk 4]
```diff
@@ -215,6 +258,11 @@ impl SchannelCredBuilder {
                 cred_data.cSupportedAlgs = supported_algorithms.len() as winapi::DWORD;
                 cred_data.palgSupportedAlgs = supported_algorithms.as_ptr() as *mut _;
             }
+            if let Some(ref enabled_protocols) = self.enabled_protocols {
+                cred_data.grbitEnabledProtocols = enabled_protocols.iter()
+                                                                   .map(|p| p.dword(direction))
+                                                                   .fold(0, |acc, p| acc ^ p);
```
Comment (sfackler) [inline src/lib.rs]: This should probably be `|` not `^` right?

File: src/lib.rs (hunk 1 above)
Comment (sfackler) [inline src/lib.rs]: You can downcast the error out - look at the bad_domain test for an example.

File: src/lib.rs (hunk 4 above)
Comment (frewsxcv) [inline src/lib.rs]: lol, yes

Comment (frewsxcv) [inline src/lib.rs]: https://github.com/sfackler/schannel-rs/pull/6/commits/84406a45a5fd04b1b96e6f6e4d46f9b227acd5f6

File: src/lib.rs (hunk 1 above)
Comment (frewsxcv) [inline src/lib.rs]: https://github.com/sfackler/schannel-rs/pull/6/commits/1bbc09b0c46b05881edd11f82c96c13cc971491f

File: src/lib.rs [hunk 5]
```diff
@@ -205,6 +242,12 @@ impl SchannelCredBuilder {
         self
     }
 
+    /// Sets the protocols enabled for sessions created from this builder.
+    pub fn enabled_protocols(mut self, enabled_protocols: &[Protocol]) -> SchannelCredBuilder {
+        self.enabled_protocols = Some(enabled_protocols.to_owned());
```
Comment (sfackler) [inline src/lib.rs:247]: You could do the fold here and avoid having to store the Vec at all.

Comment (frewsxcv) [inline src/lib.rs:247]: 👍

Comment (frewsxcv) [inline src/lib.rs:247]: We don't know what the direction is (client vs. server) at that time though

Comment (frewsxcv) [inline src/lib.rs:247]: ....unless we unconditionally set both the client _and_ server flags, which might be okay?

Comment (sfackler) [inline src/lib.rs:247]: Oh right. Let's leave it as-is for now then.

**Pull request merged**